consensus/src/main/java/org/apache/iotdb/consensus/ratis/metrics/IoTDBMetricRegistry.java

```
@@ -0,0 +1,201 @@
+
+  private final AbstractMetricService metricService;
+  private final MetricRegistryInfo info;
+  private final String prefix;
+  private final Map<String, String> metricNameCache = new ConcurrentHashMap<>();
+  private final Map<String, CounterProxy> counterCache = new ConcurrentHashMap<>();
+  private final Map<String, TimerProxy> timerCache = new ConcurrentHashMap<>();
+
+  private final Map<String, GaugeProxy> gaugeCache = new ConcurrentHashMap<>();
+
+  IoTDBMetricRegistry(MetricRegistryInfo info, AbstractMetricService service) {
+    this.info = info;
+    this.metricService = service;
+    prefix =
+        MetricRegistry.name(
+            info.getApplicationName(), info.getMetricsComponentName(), info.getPrefix());
+  }
+
+  private String getMetricName(String name) {
+    return metricNameCache.computeIfAbsent(name, n -> MetricRegistry.name(prefix, n));
+  }
+
+  @Override
+  public Timer timer(String name) {
+    final String fullName = getMetricName(name);
+    return timerCache.computeIfAbsent(
+        fullName, fn -> new TimerProxy(metricService.getOrCreateTimer(fn, MetricLevel.IMPORTANT)));
+  }
+
+  @Override
+  public Counter counter(String name) {
+    final String fullName = getMetricName(name);
+    return counterCache.computeIfAbsent(
+        fullName,
+        fn ->
+            new CounterProxy(
+                metricService.getOrCreateCounter(getMetricName(name), MetricLevel.IMPORTANT)));
+  }
+
+  @Override
+  public boolean remove(String name) {
+    // TODO (szywilliam) we can add an interface like removeTypeless(name)
+
+    try {
+      metricService.remove(MetricType.COUNTER, getMetricName(name));
+    } catch (IllegalArgumentException ignored) {
+    }
+    try {
+      metricService.remove(MetricType.TIMER, getMetricName(name));
+    } catch (IllegalArgumentException ignored) {
+    }
+    try {
+      metricService.remove(MetricType.AUTO_GAUGE, getMetricName(name));
+    } catch (IllegalArgumentException ignored) {
+    }
```

If we ignore exceptions here, why use three try?

---

To make each of them independent